uniffi_bindgen/src/interface/ffi_names.rs

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! # FFI Names
+//!
+//! This module calculates the names we export in dylibs.
+//!
+//! We have two main goals with naming things:
+//!   - Avoid name collisions.  For example when two objects having methods with the same name or two
+//!     UniFFI crates are bundled together in the same dylib.
+//!   - Guard against accidentally using foreign-language bindings generated from one version of an
+//!     interface with the compiled Rust code from a different version of that interface.  We add
+//!     checksums that depend on things like function signatures to avoid the possibility of the
+//!     bindings calling into the scaffolding when they don't agree on the interface. The result
+//!     will be an ugly inscrutable link-time error, but that is a lot better than triggering
+//!     potentially arbitrary memory unsafety!
```

Comment:
IIUC, there are some `ComponentInterface` changes that will no longer be covered by the checksum but might still lead to unsafe behaviour in the event of a version mismatch.

For example, suppose you have a function `foo(bar: Bar)` for some record type `Bar`, and you add or remove a field from `Bar`. The checksum calculated for `foo` will not change (it still takes a single argument of record type `Bar`) but its internal `Bar`-lifting code may expect to read a different set of fields than that the calling code sends in. This would *probably* result in a clean error given the strict length checks that we do in the deserialization code, but there are likely cases where it would silently do the wrong thing.

I haven't been following the proc-macro changes very closely, but I expect there's not much you can do to avoid that under the constraints of a proc-macro approach. Given just the tokens of a rust-level function signature, I don't think there's a good way to find out any of the internal details of its argument types.

It may be worth trying to enumerate any known cases that are now covered by the checksum as part of the docstring here, for completeness. It's already a best-effort attempt so that's probably good enough in practice.

Alternately, we could try to add additional guards for such cases. Some thought bubbles I've had in the past along these lines:

* The serialization format could check the checksum at runtime; for example by insisting that the first two bytes of the serialized representation of `Bar` match the compiled-time-calculated checksum of `Bar`.
* Every user-defined type `Bar` could produce a stub function `_uniffi_typecheck_<checksum>_Bar` that doesn't actually *do* anything, but that gets linked against at module load time. A linker error here would indicate that some internal details of the `Bar` type do not match the foreign code's expectations, triggering an early abort.

Reply:
This is a really good point.  In addition to those 2 solutions, I think we *might* be able to use some macro magic to include the checksum of the type definition in the scaffolding function name itself.

Reply:
I think the macro magic approach could maybe work, you'd need to do something like this:
  - Define checksum macros that input a list of other macros followed by a list of arguments a list of arguments
  - Each macro would:
     - Append their local checksum to the list of arguments
     - Prepend checksum macros for any dependent types
     - Call the first macro in the list with the arguments
  - The final macro would append the checksum to the function name
  - To define a scaffolding function, you'd put together a stack like this: `first_type_checksum!(second_type_checksum, append_checksum_to_function_name, { function def })`

So, it's possible but I don't think it's a good idea.  In addition to the complications on the macro side, if there was ever a cycle between the types then it would go into infinite recursion.

I think adding checksums to RustBuffers could present issues down the road if we ever wanted to move to using C structs or something to store those.  Also it wouldn't handle the `Object`.

Adding the stub functions and checking when we load the dylib seems like a great way to handle this to me.  Maybe they could just be a static `u8`s?

Reply:
Indeed, no need for them to be actual functions, just symbols of some sort.